crates/pipeline_manager/src/db/mod.rs

```diff
@@ -732,7 +728,7 @@ impl PipelineRevision {
 #[derive(Deserialize, Serialize, ToSchema, Eq, PartialEq, Debug, Clone)]
 pub(crate) struct PipelineDescr {
```

Comment:
In `row_to_pipeline_descr` later on when the pipeline is started and the `PipelineDescr` is constructed, it creates the `attached_connectors` field with type `Vec<AttachedConnector>`, which contains names rather than IDs. If there is a name update between the pipeline being constructed and the creation of the revision, the pipeline description in `generate_pipeline_config` can refer to a connector different from when the pipeline was created in the API call.

I think the way to get around this is to have a special `ApiAttachedConnector` that is used in `NewPipelineRequest` and `UpdatePipelineRequest`, which is resolved at API call. This will also reduce the amount of changes needed in db/mod.rs

Reply:
In this way, names are only for convenience in the API calls, and beyond that after they are translated in the API call, they are only a field in the database that happens to be unique (though the database does not use it at all, possibly only for error reporting to the user).

Reply:
Do you see this happening in actual executions? `AttachedConnector` in the DB always uses IDs, not names. We convert names to IDs and back using joins both in the write and read paths.

So:
create(pipeline), modify(connector), read(pipeline) should still reflect the updated name.

Reply:
This test for example passes:

```
#[tokio::test]
async fn update_conn_name() {
    let handle = test_setup().await;
    let tenant_id = TenantRecord::default().id;
    let connector_id = handle
        .db
        .new_connector(
            tenant_id,
            Uuid::now_v7(),
            "a",
            "b",
            &test_connector_config(),
        )
        .await
        .unwrap();
    let ac = AttachedConnector {
        name: "foo".to_string(),
        is_input: true,
        connector_name: "a".to_string(),
        relation_name: "".to_string(),
    };
    let rc = RuntimeConfig::from_yaml("");
    handle
        .db
        .new_pipeline(
            tenant_id,
            Uuid::now_v7(),
            &None,
            "1",
            "2",
            &rc,
            &Some(vec![ac]),
        )
        .await
        .unwrap();
    handle
        .db
        .update_connector(tenant_id, connector_id, "not-a", "b", &None)
        .await
        .unwrap();
    let pipeline = handle
        .db
        .get_pipeline_by_name(tenant_id, "1".to_string())
        .await
        .unwrap();
    assert_eq!(
        "not-a".to_string(),
        pipeline
            .descriptor
            .attached_connectors
            .get(0)
            .unwrap()
            .connector_name
    );
}

```

Reply:
The problem with doing name <-> id conversions from the API server side is atomicity. We can use transactions from inside the storage trait, but not from outside it.

Reply:
I meant as solution that the literal type `AttachedConnector` of  `new_pipeline` and `update_pipeline` is different from the type of `AttachedConnector` of `PipelineDescr`, in that one has a name (the former) and the other has an identifier (the latter). Name <-> id conversion happens only to what is provided by the user (through the API) in the Storage trait implementation at the start of a function there, it is part of the transaction.

What I describe will only happen in parallel execution. Assume a pipeline is created with connector id 001 with name "A". Within `pipeline_is_committable(tenant_id, pipeline_id)` the following happens:
1. `let pipeline = self get_pipeline_descr_by_id(...)` will result in a `PipelineDescr` with `AttachedConnector` having connector names resolved at that point. It will have connector id 001 with name A.
2. Now, in parallel by a different call, connector A with id 001 is renamed to B.
3. `let connectors = self.get_connectors_for_pipeline_id(...)` will now return connector 001 with name B.
4. The name of the connector in pipeline's AttachedConnector is different from that in connectors
5. This will result in problems in `PipelineRevision::validate(...)` called later

Reply:
Ah yes. I'll be reworking that code as part of a larger change in another branch (where I'll try a prototype without  the history/revisions stuff). It'll simplify this logic such that we can hopefully run this in a single transaction.